### ISSUE TITLE
Using sound repeater library

### ISSUE THREAD

**Issue by musictopia2** [AndrewAung11/Pomodoro-Clock#2]
I created a nuget package here
https://github.com/musictopia2/Reminders.WPFCoreLibrary 
that has repeaters.  Another option is I can extract the sound repeater in its own library as well.

Also, I created some libraries to make it easy to work with reminders.  The link to the basic ones are here.
https://github.com/musictopia2/Reminders.CoreLibrary 

One option I can help you with is creating new methods in the library to fit this situation.
Another option is I can create another library that uses the core but with addons to fit the situation as well.
I have done lots of reminder apps.  Would love to find a way to contribute to this.

**Comment by AndrewAung11** [comment]
I don't really understand the concept of this. Would love an explanation! Thanks

**Comment by musictopia2** [comment]
I created a complete library for doing reminders.  I made it very flexible.   If you like, I can create a very simple app that demostrates using the library.
It is intended to implement interfaces and then have different reminder processes.  By using interfaces, you can have nearly unlimited different reminder processes.
For me, I have at least 5 different reminder processes.
One is to do a bible reading at 6 am monday through friday.  I have some weekly reminders as well.
Plus one that is the friday after thanksgiving every year.
Since interfaces does not care how it gets it, then all it cares about is figuring out when the reminder is.  Once that reminder happens, then uses the wpf part for the sound repeater.  I mentioned that i could have extracted it to its own library.  i have not yet though since the primary purpose of sound repeaters is for reminders.

**Comment by AndrewAung11** [comment]
This project is made with WinForms so I don't think the WPF library would work. However, if you have a library for WinForms, then yes, I would love to see a demonstration please.

**Comment by musictopia2** [comment]
I can see if I can make it work with windows forms.  Does windows forms have any media playing capabilities?  because the key is having something that can play the media which is the sounds that repeat.

**Comment by AndrewAung11** [comment]
For this project, I have used [System.Media.SoundPlayer](https://learn.microsoft.com/en-us/dotnet/api/system.media.soundplayer?view=windowsdesktop-7.0)
You can see the code on [line 18 of PomodoClock.cs file](https://github.com/AndrewAung11/Pomodoro-Clock/blob/main/PomodoroClock/PomodoClock.cs#L18).

**Comment by musictopia2** [comment]
I will create a sample for you.  Since I don't use windows forms much, then i would create as simple of a sample as possible.
Would you be open to using blazor with windows forms?  Because there is an option for doing blazor with windows forms as well.
One possibility can be the sample would use blazor with windows forms.  However, you can replace with pure windows forms as well.  The business logic stays the same regardless of ui used.

**Comment by AndrewAung11** [comment]
I have done a lot of HTML/CSS in past but I am not willing to use blazor for this project. However, if you would make the sample with blazor and WinForms, I could try to understand what is written.

**Comment by musictopia2** [comment]
The blazor part of it would be super simple.

**Comment by musictopia2** [comment]
One major thing I will do as well to make the sound repeater usable to more people is that part would be its own library for the windows version.  I found I was using the soundplayer for my repeater.  However, the changes will not even depend on blazor so if somebody somehow wanted to use without blazor, they can.  Of course, one interface you will have to implement is an IPopup which you have to decide how the form that implements it looks like.  The core code will call into it not knowing whether its windows forms, wpf or even mobile.

**Comment by musictopia2** [comment]
One question I have is do you know how to use dependency injection in windows forms?  If not, are you willing to learn it?
Because the reminder processes reply heavily on dependency injection.  Makes it a much more flexible system so you have many choices of what to use to implement the interfaces and have it more loosely coupled.

**Comment by AndrewAung11** [comment]
I don't know what it is. Could you dumb it down for me?

**Comment by musictopia2** [comment]
Here is a video that explains dependency injection.
https://www.youtube.com/watch?v=Hhpq7oYcpGE 
My sample will show using dependency injection in windows forms as well.

**Comment by musictopia2** [comment]
One question.  Would you like to use .net 7 which officially releases november 8 or stick with .net 6?
Because if so, then I can update all my libraries to use .net 7.
Otherwise, my major changes will first be .net 6 then later .net 7.

**Comment by musictopia2** [comment]
I just created a new repository here
https://github.com/musictopia2/Reminders.WindowsSoundRepeaterLibrary 
which has the sound repeater alone.  Once I find out whether you want to use .net 7 or stick with .net 6, then I can continue work on creating the samples.  I will create one that uses just the repeater alone and the other that integrates into my reminder processes.
The advantage of using my reminder processes is there is lots of code that is already done that makes it easy to create any type of reminder (even though you have some setup to do).  If you go with the simple option, then the library alone will allow you to have the reminders but you have to think about how to do the rest including the popups.  If you prefer just the sample for the repeater but not the entire system, that is fine too.  The more complex sample would mean you have to implement an interface which represents a popup when the reminder happens.  By you implementing, you can design the screen any way you want.  I have screens created but it uses blazor since i love blazor.

**Comment by AndrewAung11** [comment]
I am using .net 4 for this project. Also, just to let you know before it gets too late, I have completed nearly everything for this project. Everything works well and the only thing left is to polish the UI.

**Comment by musictopia2** [comment]
I can change to use .net standard for the repeater.  I hope you are at least using a version of .net that is compatible with .net standard.
Another choice you have is to copy/paste from my repository into your library as well.
Since you are mostly done, then it would probably be best to use the sound repeater process without the parts for the reminder core processes since you probably did your own version of that part.  the sound repeater only focuses on one class that has 2 methods.   One is to start it and decide how often to repeat and the other is to turn off.

**Comment by AndrewAung11** [comment]
Yes, I am pretty sure I am using a version of .net that is compatible with .net standard.

**Comment by musictopia2** [comment]
The bad news is when I try to make it use .net standard, then it fails to compile because it relies on some extensions that only work with .net 6 or .net 7.  Tried some other things.  Unfortunately, since it relies on windows stuff for playing the media, the best I can do is you can fork the repository if you want and you can copy/paste all of it to use on yours.
If you were willing to upgrade to at least .net 6, then the library can be used.

**Comment by AndrewAung11** [comment]
I would like to, but I do not know how.

**Comment by musictopia2** [comment]
I can go ahead and fork and make the changes to upgrade to .net 6.  I know I would be able to upgrade to .net 6 with no problem.  If I have your permission, then I will go ahead and fork and step one will be to do whatever is necessary to make it work with .net 6.
.net 6 for sure works with windows forms.  Later today, it will work with .net 7.  If you want, I can even target .net 7 since it officially releases today.

**Comment by AndrewAung11** [comment]
Yeah, no need my permission. All contributions are welcome. Also you can choose any version of the .net you like, just choose a stable one. The code I have written are all generic C# codes, so you might not need to change that much. Only maybe a few for IDE auto-created codes.

**Comment by musictopia2** [comment]
The .net 7 being released today is the stable one.  Its the supported version for 18 months.
The version before today is release candidate 2.  In a few hours from now is the stable version brand new for .net 7.  So I will go ahead and use that one.

**Comment by AndrewAung11** [comment]
Unfortunately, I couldn't get it running. I took last few weeks trying to figure out what to do but couldn't come up with a solution other than make some big changes which is a lot of work to do.
Also, as this project has been taking too much time, I have decided to just finish it as it is now and not make it advanced. Sorry for your time that I have wasted.

**Comment by musictopia2** [comment]
Did you have .net 7 installed on your computer?  I do know it ran perfectly on my side.   If you installed the latest for visual studio 2022 with update 4, then .net 7 would already be installed.  It was not a waste anyways.  Did not take me that long to do the work anyways.   If you mentioned to me about your problems, I could have helped you as well.

**Comment by AndrewAung11** [comment]
yeah, the thing is I am using visual studio 2019 and as you know how the process go, it is gonna take a long time uninstalling the one I have right now and then installing vs 2022 again. I am also still a student attending school, so yeah.
